hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java

```diff
@@ -255,14 +345,28 @@ public int compare(FileSummary.Section s1, FileSummary.Section s2) {
         case INODE: {
           currentStep = new Step(StepType.INODES);
           prog.beginStep(Phase.LOADING_FSIMAGE, currentStep);
-          inodeLoader.loadINodeSection(in, prog, currentStep);
+          stageSubSections = getSubSectionsOfName(
+              subSections, SectionName.INODE_SUB);
+          if (loadInParallel && (stageSubSections.size() > 0)) {
+            inodeLoader.loadINodeSectionInParallel(executorService,
+                stageSubSections, summary.getCodec(), prog, currentStep);
+          } else {
+            inodeLoader.loadINodeSection(in, prog, currentStep);
+          }
         }
           break;
         case INODE_REFERENCE:
           snapshotLoader.loadINodeReferenceSection(in);
           break;
         case INODE_DIR:
-          inodeLoader.loadINodeDirectorySection(in);
+          stageSubSections = getSubSectionsOfName(
+              subSections, SectionName.INODE_DIR_SUB);
+          if (loadInParallel && stageSubSections.size() > 0) {
```

Comment:
Would you like to add some unit test to cover serial loading and parallel loading for old&new format fsimage?

Reply:
Yes, I need to figure out how to add some tests for this overall change.

Reply:
you probably need to put a sample fsimage in old format in the resource directory (hadoop-hdfs-project/hadoop-hdfs/src/test/resources/),  and load it in a unit test

Reply:
I have a unit test which ensures a parallel image can be created and loaded. It would be fairly easy to create another test which generates a non-parallel image, validates it is non-parallel and then attempt to load it with parallel enabled. Do you think that would cover what we need?
I also have a test that enables parallel and compression and then it verifies the parallel part is not used as compression disables it.